server/coordinator/procedure/manager_test.go

```diff
@@ -0,0 +1,149 @@
+	return m.state
+}
+
+func (m *MockProcedure) RelatedVersionInfo() procedure.RelatedVersionInfo {
+	return m.relatedVersionInfo
+}
+
+func (m *MockProcedure) Priority() procedure.Priority {
+	return procedure.PriorityMed
+}
+
+func TestManager(t *testing.T) {
+	ctx := context.Background()
+	re := require.New(t)
+
+	c := test.InitStableCluster(ctx, t)
+	manager, err := procedure.NewManagerImpl(zap.NewNop(), c.GetMetadata())
+	re.NoError(err)
+
+	err = manager.Start(ctx)
+	re.NoError(err)
+
+	procedureID := uint64(0)
+	// Test submit multi single shard procedure.
+	snapshot := c.GetMetadata().GetClusterSnapshot()
+	for shardID, shardView := range snapshot.Topology.ShardViewsMapping {
+		err = manager.Submit(ctx, &MockProcedure{
+			id:                 procedureID,
+			state:              procedure.StateInit,
+			relatedVersionInfo: procedure.RelatedVersionInfo{ClusterID: c.GetMetadata().GetClusterID(), ShardWithVersion: map[storage.ShardID]uint64{shardID: shardView.Version}, ClusterVersion: c.GetMetadata().GetClusterViewVersion()},
+			execTime:           time.Millisecond * 50,
+		})
+		procedureID++
+		re.NoError(err)
+	}
+	// Procedures state is init, list result length should be 0.
+	infos, err := manager.ListRunningProcedure(ctx)
+	re.NoError(err)
+	re.Equal(0, len(infos))
+	time.Sleep(time.Millisecond * 10)
+	// Procedures state is running, list result length should be len(shardViews).
+	infos, err = manager.ListRunningProcedure(ctx)
+	re.NoError(err)
+	re.Equal(len(snapshot.Topology.ShardViewsMapping), len(infos))
+	time.Sleep(time.Millisecond * 100)
+	// Procedures state is finish, list result length should be 0.
+	infos, err = manager.ListRunningProcedure(ctx)
+	re.NoError(err)
+	re.Equal(0, len(infos))
+
+	// Test submit procedure with multi shards.
+	shardWithVersions := map[storage.ShardID]uint64{}
+	for id, view := range snapshot.Topology.ShardViewsMapping {
+		shardWithVersions[id] = view.Version
+	}
+	err = manager.Submit(ctx, &MockProcedure{
+		id:                 procedureID,
+		state:              procedure.StateInit,
+		relatedVersionInfo: procedure.RelatedVersionInfo{ClusterID: c.GetMetadata().GetClusterID(), ShardWithVersion: shardWithVersions, ClusterVersion: c.GetMetadata().GetClusterViewVersion()},
+		execTime:           time.Millisecond * 100,
+	})
+	re.NoError(err)
+	procedureID++
+	for shardID, shardView := range snapshot.Topology.ShardViewsMapping {
+		err = manager.Submit(ctx, &MockProcedure{
+			id:                 procedureID,
+			state:              procedure.StateInit,
+			relatedVersionInfo: procedure.RelatedVersionInfo{ClusterID: c.GetMetadata().GetClusterID(), ShardWithVersion: map[storage.ShardID]uint64{shardID: shardView.Version}, ClusterVersion: c.GetMetadata().GetClusterViewVersion()},
+			execTime:           time.Millisecond * 50,
+		})
+		procedureID++
+		re.NoError(err)
+	}
+	// Procedure with multi shard is running, all shard is locked.
+	time.Sleep(time.Millisecond * 10)
```

Comment:
Do it with a loop.

Reply:
The unit tests here are very time sensitive and cannot be directly replaced by loops.